File: Day<28>/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    int min_sprinklers(int gallery[], int n)
    {
        vector<pair<int,int>> sprinklers;
        for (int i=0; i<n; i++)
            if (-1 < gallery[i])
                sprinklers.push_back( pair<int,int> ( i-gallery[i], i+gallery[i] ) );
        
        sort(sprinklers.begin(), sprinklers.end());
        
        int target=0, sprinklers_on=0, i=0;
        while (target < n)
        {
            if (i==sprinklers.size() || target<sprinklers[i].first)
                return -1;
            
            int max_range = sprinklers[i].second;
            while( i+1<sprinklers.size() && sprinklers[i+1].first<=target )
            {
                i++;
                max_range = max( max_range,  sprinklers[i].second );
            }
            if (max_range < target)
                return -1;
            
            sprinklers_on++;
            target = max_range + 1;
            i++;
        }
        return sprinklers_on;
    }
};

int main()
{
    int t;
    cin>>t;
    while (t--)
    {
        int n;
        cin>>n;
        int gallery[n];
        for (int i=0; i<n; i++)
            cin>> gallery[i];
        Solution obj;
        cout << obj.min_sprinklers(gallery,n) << endl;
    }
	return 1;
}

File: Day<28>/Solution.py
class Solution:
    def min_sprinklers(self, gallery, n):
        sprinklers = []
        for i in range(n):
            if -1<gallery[i]:
                sprinklers.append( [i-gallery[i] , i+gallery[i]] );
        sprinklers.sort()
        target=0
        sprinklers_on=0
        i=0
        while(target<n):
            if i==len(sprinklers) or target<sprinklers[i][0]:
                return -1
            max_range = sprinklers[i][1]
            while( i+1<len(sprinklers) and sprinklers[i+1][0]<=target ):
                i+=1
                max_range = max(max_range,sprinklers[i][1])
            if max_range<target:
                return -1
            sprinklers_on+=1
            target=max_range+1
            i+=1
        return sprinklers_on


if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input())
        gallery = [int(x) for x in input().strip().split()]
        obj = Solution()
        print(obj.min_sprinklers(gallery,n))
